pkg/terraform/controller.go

```diff
@@ -0,0 +1,233 @@
+
+	rl := ratelimiter.NewDefaultProviderRateLimiter(ratelimiter.DefaultProviderRPS)
+	o := controller.Options{
+		RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+	}
+
+	r := managed.NewReconciler(mgr,
+		xpresource.ManagedKind(of),
+		managed.WithInitializers(),
+		managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), providerConfig: pcFn, logger: l}),
+		managed.WithLogger(l.WithValues("controller", name)),
+		managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))))
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(o).
+		For(obj).
+		Complete(r)
+}
+
+type connector struct {
+	kube           client.Client
+	providerConfig ProviderConfigFn
+	logger         logging.Logger
+}
+
+func (c *connector) Connect(ctx context.Context, mg xpresource.Managed) (managed.ExternalClient, error) {
+
+	// TODO(hasan): create and pass the implementation of tfcli builder once available
+
+	/*	tr, ok := mg.(resource.Terraformed)
+		if !ok {
+			return nil, errors.New(errUnexpectedObject)
+		}
+		pc, err := c.providerConfig(ctx, c.kube, mg)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get provider config")
+		}
+		tfcb := tfcli.NewClientBuilder().
+			WithLogger(c.logger).
+			WithResourceName(tr.GetName()).
+			WithHandle(string(tr.GetUID())).
+			WithProviderConfiguration(pc).
+			WithResourceType(tr.GetTerraformResourceType())*/
+
+	return &external{
+		kube:   c.kube,
+		tf:     conversion.NewCli(nil),
+		log:    c.logger,
+		record: event.NewNopRecorder(),
+	}, nil
+}
+
+type external struct {
+	kube client.Client
+	tf   conversion.Adapter
+
+	log    logging.Logger
+	record event.Recorder
+}
+
+func (e *external) Observe(ctx context.Context, mg xpresource.Managed) (managed.ExternalObservation, error) {
+	tr, ok := mg.(resource.Terraformed)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errUnexpectedObject)
+	}
+
+	if xpmeta.GetExternalName(tr) == "" && meta.GetState(tr) == "" {
+		return managed.ExternalObservation{
+			ResourceExists: false,
+		}, nil
+	}
+
+	res, err := e.tf.Observe(tr)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, "cannot check if resource exists")
+	}
+
+	if !res.Completed {
+		// Note(hasan): If an async operation is not completed yet, we don't want to reconcile
+		// with exponential backoff since this is not an error condition rather an expected situation.
+		// And checking that frequently wouldn't make much sense since we expect terraform client
+		// invokes to take some seconds. Here, we chose to check again after poll interval,
+		// but to speed things up, we might want to consider adding a special wait that is less frequent
+		// than exponential backoff but more frequent than poll interval.
+
+		// Observation is in progress, do nothing. We will check again after the poll interval.
+		return managed.ExternalObservation{
+			ResourceExists:   true,
```

Comment:
As you've noticed, this (your first comment) is exactly the case already. `Adapter` interface is exactly responsible for hiding the underlying logic related to cli implementation from Crossplane managed reconciler by doing conversions (tfstate to parameters) and returning appropriate observations.

From cli point of view, it makes sense to return an error when a new operation is requested while another one is in progress (e.g. Trying to trigger an `Observe` while a `Create` is in progress). At this point, tfcli implementation of `Adapter` interface gracefully handles this error and returns a proper observation result. It could be said to be expected the logic not to be straightforward at first look given the purpose of `Adapter` is hiding underlying logic.

I think only making `Observe` blocking wouldn't help too much unless we start allowing multiple operations to run in parallel on a single resource (at least being able to do an Observation while Create/Update/Delete is in progress). Assume we did a blocking `Observe`, noticed resource is not up to date and started a non-blocking `Update`. In the next reconciliation, if Update is still in progress, how could we start a new `Observe`? 

I would love to hear @ulucinar's thoughts on this as this discussion is mostly related to the interface provided by tfcli package.

Reply:
Sorry I believe three comments made it harder to make my point. I was meaning to make the actual `terraform refresh` call blocking. For example, when `Observe` called and another operation like Create/Update/Delete is in progress, we'd return saying it's in progress. But if no operation is in progress, then the `terraform refresh` call we make should be blocking. That would reduce the amount of states we need to think if I understood the structure correctly.

Reply:
As I commented below, we can introduce a blocking `Observe` and continue blocking concurrent `tfcli` operations.

Reply:
Updated accordingly, as we discussed offline 👍